### ISSUE TITLE
Add bitwise operators

### ISSUE THREAD

**Issue by drbo** [LuaJIT/LuaJIT#63]
Lua 5.3 supports bitwise operators.  LuaJIT already has bitwise functions in the `bit` library, so adding the operators should be a matter of lexer/parser support.

**Comment by fsfod** [comment]
You would also need to add new bytecode instructions for them as well.  Unless you lowered them to direct calls to the bitwise functions which would need to be added to the constant list of the functions proto during parsing

**Comment by daurnimator** [comment]
Note that in Lua 5.3 the bitwise operators work on 64bit integers (which are currently only available in luajit via the ffi!).

**Comment by naoliv** [comment]
:+1:

**Comment by fogfx** [comment]
@drbo of course, you do run into the problem of people seeing these and assuming they behave like they do in 5.3... do they?

For example 5.3 guarantees that, for shifts:
- a shift operation returns 0 if the shift count is too large. Bitop only uses the lower 5 (or 6) bits of the shift count and thus does not guarantee this.
- negative shift counts are allowed and shift in the opposite direction. I recall Mike was not too impressed about this back in 5.2 ;). As implied above, bitop doesn't support this either.

There are also other considerations like non-integral floats and out of range floats being used in bit operations; both of those throw errors in 5.3. The former is UB for bitop but the latter is defined behavior only for numbers within [-2^51, 2^51] iirc; either way bitop will not throw. I'm sure there are other differences. And 5.3's semantics don't come for free, of course.

So... a decision would have to be made whether these are 5.3-compliant or are just (possibly metamethod triggering) sugar over bitop, which is the impression I get from the OP.

In theory, the operators could use 5.3 semantics while bitop serves a more low-level efficiency role (ie basically raw operations), but those semantic differences could still be rather unpleasant, especially if people port their bitop-based code thinking the operators must inherently be faster ;) (maybe interpreted, sure)

**Comment by Hydroque** [comment]
What is the progress with this?

**Comment by MikePall** [comment]
Moving the discussion from PR #312 back here:

The LuaJIT VM doesn't use the numeric type model of Lua 5.3 and doesn't use its mis-designed and slow bit operation semantics, either. So the only reasonable choice is for these operators to be simple shorthands for the existing LuaJIT `bit.*()` functions and their semantics. Which means, they need to support FFI types, too -- probably in the fallback path for the interpreter, but easily JIT-compiled.

Also, since it's a new feature, the troublesome implicit string-to-number coercion should not be allowed.

There's the dreaded question what operator to use for logical vs. arithmetic right shift: even though LuaJIT (non-FFI) bit operations work on signed numbers, I guess using `>>` for logical and `>>>` for arithmetic right shift makes more sense. Note that Java and Javascript have this the other way round. I strongly disagree with Lua 5.3 completely omitting arithmetic right shift.

I don't think we'd want or need the full range of VV/VN/NV bytecode variants for any of these operators. One VV bytecode for each operator would probably suffice.

Alternatively, we could instead add a generic intrinsic call instruction to the bytecode: RA = dest, RB = rbase, RC = intrinsic ID. The two operands are addressed by rbase and need to be next to each other, just like `CAT` or calls. Which works out most of the time in nested expressions, I guess. Slot copies are not that costly and half of the operands would need an extra `KSHORT` or `KNUM`, anyway. No function object is needed -- the intrinsic must behave like an operator.

I'm sure that such an intrinsic call bytecode would simplify the code a lot.

[And it's generally useful: e.g. rewriting the `package` module in pure Lua would be possible then. There are other higher-level functions that are currently fully implemented in C, but could be rewritten in Lua with just a few common low-level intrinsics. It's much easier to JIT-compile the fast paths of the surrounding Lua code plus those intrinsics, rather than emulate the semantics of a complex piece of C code in the JIT compiler frontend.]

**Comment by SoniEx2** [comment]
At least Lua doesn't surprise you by making 1 >> 32 == 0 on a build with 64-bit integers but 1 on a build with 32-bit integers. (But you *could* argue it should just error.)

**Comment by creationix** [comment]
I also eagerly await this.  For now I've made some quick C functions that FFI can call: https://github.com/luvit/luvi/commit/5af13b7b05be955ba303a80a667ac7cbdbea1d41

**Comment by Lejo1** [comment]
Any Progress on this or lua 5.3-support in general?

**Comment by slandden** [comment]
@Lejo1 lua 5.3 is a *different* language.

**Comment by stuta** [comment]
This would make converting C code to Luajit much easier. This is the only thing I miss from other Lua versions.

**Comment by ccuser44** [comment]
I'm really not sure if this should be added. It would complicate the compiler and add unnecessary methamethods.

If internally it would use the `bit.` library then you could just use the `bit` library instead. The bitwise operators only allow a subset of the operations the `bit.` library can do anyways.

**Comment by slandden** [comment]
I know you  are, but what am I?

21 ene 2023, 06:26 por ***@***.***:

>
>
>
>
> I'm really not sure if this should be added. It would complicate the compiler and add unnecessary methamethods.
>
>
You are the most intelligent person in the world. Everybody wants to listen to your opinion.

**Comment by slandden** [comment]
And regarding the username: ​Nobody cares that you are gay Noam Chimsky. The reason you put me in a gay conversion prison is because you are a slave.
You are so racist that you turned the racist terms (which the refugee agency in Georgia did not understand, despite the translator having a solid command of English) into a literal semantic interpretation, and declared yourself a Georgian the same way as Nation of Islam declaring themselves Moors.
But you are not a Georgian, and addressing me by my transliterated Georgian name when holding a lynch mob hearing about Vladimir Putin prooves this.

Peace,

Shawn Paul Landden
21 ene 2023, 11:16 por ***@***.***:

**Comment by MikePall** [comment]
Closing here in favor of:
Discussion about the syntax: #1475
Intent to backport to v2.1: #1476.
